### PR TITLE
[Misc] Support attention logits soft-capping with flash-attn

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR adds support for attention logits soft-capping in the FlashAttention backend.

This is done by
1. Moving `logits_soft_cap` from `AttentionMetadata` to `AttentionImpl`.
2. Using `vllm-flash-attn == 2.6.1` which added support for soft-capping.

Note that vllm-flash-attn v2.6.1 uses PyTorch 2.4.0, so the PR must be merged after #6951

This will hopefully resolve many of the issues in running Gemma2 models.

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
LGTM
